### PR TITLE
Add permission based functionality

### PR DESCRIPTION
**Pull request by charithag**
- Add permission based functionality
- Add Enroll device button to device details page

### CODE REVIEW

File: features/cdmf-base-ui/org.wso2.carbon.devicemgt.ui.feature/src/main/resources/jaggeryapps/devicemgt/app/units/cdmf.unit.device.listing/listing.js [hunk 1]
```diff
@@ -1,16 +1,41 @@
...
-        permissions.push("LIST_DEVICES");
-    }else if(userModule.isAuthorized("/permission/admin/device-mgt/user/devices/list")){
-        permissions.push("LIST_OWN_DEVICES");
-    }else if(userModule.isAuthorized("/permission/admin/device-mgt/emm-admin/policies/list")){
-        permissions.push("LIST_POLICIES");
-    }
     var currentUser = session.get(constants.USER_SESSION_KEY);
-    context.permissions = stringify(permissions);
-    context.currentUser = currentUser;
+    if (currentUser){
```
Comment (cnapagoda) [inline features/cdmf-base-ui/org.wso2.carbon.devicemgt.ui.feature/src/main/resources/jaggeryapps/devicemgt/app/units/cdmf.unit.device.listing/listing.js:24]: multiple formatting issues

File: features/cdmf-base-ui/org.wso2.carbon.devicemgt.ui.feature/src/main/resources/jaggeryapps/devicemgt/app/units/cdmf.unit.ui.navbar.nav-menu/nav-menu.js [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+function onRequest(context) {
+    var userModule = require("/modules/user.js").userModule;
+    var constants = require("/modules/constants.js");
+    var carbonUser = session.get(constants.USER_SESSION_KEY);
+    if (carbonUser){
```
Comment (cnapagoda) [inline features/cdmf-base-ui/org.wso2.carbon.devicemgt.ui.feature/src/main/resources/jaggeryapps/devicemgt/app/units/cdmf.unit.ui.navbar.nav-menu/nav-menu.js:23]: formatting issue

File: features/cdmf-base-ui/org.wso2.carbon.devicemgt.ui.feature/src/main/resources/jaggeryapps/devicemgt/modules/device.js [hunk 3]
```diff
@@ -332,5 +334,75 @@ deviceModule = function () {
         return license;
     };
 
+    publicMethods.getOwnDevices = function () {
+        devicemgtService = devicemgtProps["httpsURL"] + "/common/device_manager";
+        listAllDevicesEndPoint = devicemgtService + "/device/user/" + user.username + "/all";
+        var result = get(listAllDevicesEndPoint, {}, "json");
+        var devices = result.data;
+        var device;
+        for (var d in devices){
```
Comment (cnapagoda) [inline features/cdmf-base-ui/org.wso2.carbon.devicemgt.ui.feature/src/main/resources/jaggeryapps/devicemgt/modules/device.js:343]: formatting issue

File: features/cdmf-base-ui/org.wso2.carbon.devicemgt.ui.feature/src/main/resources/jaggeryapps/devicemgt/modules/device.js [hunk 4]
```diff
@@ -332,5 +334,75 @@ deviceModule = function () {
...
+        var allDevices = [];
+        var deviceCount = unGroupedDevices.length;
+        for (var g in user_groups) {
+            var deviceInGroup = user_groups[g].devices;
+            deviceCount += deviceInGroup.length;
+            if (deviceInGroup && deviceInGroup.length == 0) {
+                delete user_groups[g]["devices"];
+            }
+            var device;
+            for (var d in deviceInGroup){
```
Comment (cnapagoda) [inline features/cdmf-base-ui/org.wso2.carbon.devicemgt.ui.feature/src/main/resources/jaggeryapps/devicemgt/modules/device.js:390]: formatting issue

File: features/cdmf-base-ui/org.wso2.carbon.devicemgt.ui.feature/src/main/resources/jaggeryapps/devicemgt/modules/group.js [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+    groupModule.getDevices = function (groupId) {
+        var result = groupModule.getGroup(groupId);
+        var group = result.data;
+        if (group) {
+            //URL: GET https://localhost:9443/devicecloud/group_manager/group/id/{groupId}/device/all
+            endPoint = deviceCloudService + "/group/id/" + groupId + "/device/all";
+            data = {"username": user.username};
+            result = get(endPoint, data, "json");
+            var devices = result.data;
+            for (var d in devices){
```
Comment (cnapagoda) [inline features/cdmf-base-ui/org.wso2.carbon.devicemgt.ui.feature/src/main/resources/jaggeryapps/devicemgt/modules/group.js:177]: formatting issue

File: features/cdmf-base-ui/org.wso2.carbon.devicemgt.ui.feature/src/main/resources/jaggeryapps/devicemgt/modules/user.js [hunk 6]
```diff
@@ -180,6 +180,15 @@ var userModule = function () {
             } else {
                 var defaultUserClaims = privateMethods.buildDefaultUserClaims(firstname, lastname, emailAddress);
 
+                for (var role in userRoles){
```
Comment (cnapagoda) [inline features/cdmf-base-ui/org.wso2.carbon.devicemgt.ui.feature/src/main/resources/jaggeryapps/devicemgt/modules/user.js:183]: multiple formatting issues

File: features/cdmf-base-ui/org.wso2.carbon.devicemgt.ui.feature/src/main/resources/jaggeryapps/devicemgt/app/units/cdmf.unit.device.listing/listing.js (hunk 1 above)
Comment (charithag) [reply features/cdmf-base-ui/org.wso2.carbon.devicemgt.ui.feature/src/main/resources/jaggeryapps/devicemgt/app/units/cdmf.unit.device.listing/listing.js:24]: Acknowledge, Already fixed in moved code with commit https://github.com/wso2/carbon-device-mgt/commit/e537e2d841ca3c7946cd34e9da96f261e0ac94e1

File: features/cdmf-base-ui/org.wso2.carbon.devicemgt.ui.feature/src/main/resources/jaggeryapps/devicemgt/app/units/cdmf.unit.ui.navbar.nav-menu/nav-menu.js (hunk 2 above)
Comment (charithag) [reply features/cdmf-base-ui/org.wso2.carbon.devicemgt.ui.feature/src/main/resources/jaggeryapps/devicemgt/app/units/cdmf.unit.ui.navbar.nav-menu/nav-menu.js:23]: Acknowledge, Already fixed in moved code with commit https://github.com/wso2/carbon-device-mgt/commit/e537e2d841ca3c7946cd34e9da96f261e0ac94e1

File: features/cdmf-base-ui/org.wso2.carbon.devicemgt.ui.feature/src/main/resources/jaggeryapps/devicemgt/modules/device.js (hunk 3 above)
Comment (charithag) [reply features/cdmf-base-ui/org.wso2.carbon.devicemgt.ui.feature/src/main/resources/jaggeryapps/devicemgt/modules/device.js:343]: Acknowledge, Already fixed in moved code with commit https://github.com/wso2/carbon-device-mgt/commit/e537e2d841ca3c7946cd34e9da96f261e0ac94e1

File: features/cdmf-base-ui/org.wso2.carbon.devicemgt.ui.feature/src/main/resources/jaggeryapps/devicemgt/modules/device.js (hunk 4 above)
Comment (charithag) [reply features/cdmf-base-ui/org.wso2.carbon.devicemgt.ui.feature/src/main/resources/jaggeryapps/devicemgt/modules/device.js:390]: Acknowledge, Already fixed in moved code with commit https://github.com/wso2/carbon-device-mgt/commit/e537e2d841ca3c7946cd34e9da96f261e0ac94e1

File: features/cdmf-base-ui/org.wso2.carbon.devicemgt.ui.feature/src/main/resources/jaggeryapps/devicemgt/modules/group.js (hunk 5 above)
Comment (charithag) [reply features/cdmf-base-ui/org.wso2.carbon.devicemgt.ui.feature/src/main/resources/jaggeryapps/devicemgt/modules/group.js:177]: Acknowledge, Already fixed in moved code with commit https://github.com/wso2/carbon-device-mgt/commit/e537e2d841ca3c7946cd34e9da96f261e0ac94e1

File: features/cdmf-base-ui/org.wso2.carbon.devicemgt.ui.feature/src/main/resources/jaggeryapps/devicemgt/modules/user.js (hunk 6 above)
Comment (charithag) [reply features/cdmf-base-ui/org.wso2.carbon.devicemgt.ui.feature/src/main/resources/jaggeryapps/devicemgt/modules/user.js:183]: Acknowledge, Already fixed in moved code with commit https://github.com/wso2/carbon-device-mgt/commit/e537e2d841ca3c7946cd34e9da96f261e0ac94e1

**Pull request merged**